okta/client.go

```diff
@@ -0,0 +1,204 @@
+	conf       OktaConfig
+	uspClient  *uspclient.Client
+	httpClient *http.Client
+
+	chStopped chan struct{}
+	wgSenders sync.WaitGroup
+	doStop    *utils.Event
+
+	ctx context.Context
+}
+
+type OktaConfig struct {
+	ClientOptions uspclient.ClientOptions `json:"client_options" yaml:"client_options"`
+	ApiKey        string                  `json:"apikey" yaml:"apikey"`
+	URL           string                  `json:"url" yaml:"url"`
+}
+
+func (c *OktaConfig) Validate() error {
+	if err := c.ClientOptions.Validate(); err != nil {
+		return fmt.Errorf("client_options: %v", err)
+	}
+	if c.URL == "" {
+		return errors.New("missing url")
+	}
+	if c.ApiKey == "" {
+		return errors.New("missing api key")
+	}
+	return nil
+}
+
+func NewOktaAdapter(conf OktaConfig) (*OktaAdapter, chan struct{}, error) {
+	var err error
+	a := &OktaAdapter{
+		conf:   conf,
+		ctx:    context.Background(),
+		doStop: utils.NewEvent(),
+	}
+
+	a.uspClient, err = uspclient.NewClient(conf.ClientOptions)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	a.httpClient = &http.Client{
+		Timeout: 30 * time.Second,
+		Transport: &http.Transport{
+			Dial: (&net.Dialer{
+				Timeout: 10 * time.Second,
+			}).Dial,
+		},
+	}
+
+	a.chStopped = make(chan struct{})
+
+	a.wgSenders.Add(1)
+	go a.fetchEvents(logsURL)
+
+	go func() {
+		a.wgSenders.Wait()
+		close(a.chStopped)
+	}()
+
+	return a, a.chStopped, nil
+}
+
+func (a *OktaAdapter) Close() error {
+	a.conf.ClientOptions.DebugLog("closing")
+	a.doStop.Set()
+	a.wgSenders.Wait()
+	err1 := a.uspClient.Drain(1 * time.Minute)
+	_, err2 := a.uspClient.Close()
+	a.httpClient.CloseIdleConnections()
+
+	if err1 != nil {
+		return err1
+	}
+
+	return err2
+}
+
+func (a *OktaAdapter) fetchEvents(url string) {
+	defer a.wgSenders.Done()
+	defer a.conf.ClientOptions.DebugLog(fmt.Sprintf("fetching of %s events exiting", url))
+
+	for !a.doStop.WaitFor(30 * time.Second) {
+		// The makeOneRequest function handles error
+		// handling and fatal error handling.
+		items := a.makeOneRequest(url)
+		if items == nil {
+			continue
+		}
+
+		for _, item := range items {
+			msg := &protocol.DataMessage{
+				JsonPayload: item,
+				TimestampMs: uint64(time.Now().UnixNano() / int64(time.Millisecond)),
+			}
+			if err := a.uspClient.Ship(msg, 10*time.Second); err != nil {
+				if err == uspclient.ErrorBufferFull {
+					a.conf.ClientOptions.OnWarning("stream falling behind")
+					err = a.uspClient.Ship(msg, 1*time.Hour)
+				}
+				if err == nil {
+					continue
+				}
+				a.conf.ClientOptions.OnError(fmt.Errorf("Ship(): %v", err))
+				a.doStop.Set()
+				return
+			}
+		}
+	}
+}
+
+func (a *OktaAdapter) makeOneRequest(url string) []utils.Dict {
+
+	// Prepare the request body.
+	reqData := opRequest{}
+	b, err := json.Marshal(reqData)
+	if err != nil {
+		a.doStop.Set()
+		return nil
+	}
+
+	// Get request timestamp
+	currentTime := time.Now().Unix()
+	thirtySecondsAgo := time.Unix(currentTime-120, 0).UTC().Format(time.RFC3339)
```

Comment:
Is there a way we can parse the last timestamp of the last event we got from their API and use that as the start timestamp for the next request. This will avoid possible race conditions where an event occurs in the cracks between that 30s ago and the end of the previous query and we'd miss those events.

Using the last TS of the last event fetched means we won't miss any events.